### PR TITLE
Add cluster test framework doc.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Problem

Book is lacking a doc for recommended approach to testing the cluster. 

#### Summary of Changes

These changes cover the current implementation.

Fixes #

### CODE REVIEW

File: book/src/SUMMARY.md [hunk 1]
```diff
@@ -32,6 +32,8 @@
   - [JSON RPC API](jsonrpc-api.md)
   - [JavaScript API](javascript-api.md)
   - [solana-wallet CLI](wallet.md)
+- [Testing](testing.md)
```
Comment (garious) [inline book/src/SUMMARY.md]: Once implemented, it'd be fine to include this in the `API Reference` section.

File: book/src/SUMMARY.md [hunk 2]
```diff
@@ -32,6 +32,8 @@
...
+  - [Cluster Test Framework](cluster-test-framework.md)
```
Comment (garious) [inline book/src/SUMMARY.md]: Per CONTRIBUTING.md, this should sit in the proposed changes section until implemented?

File: book/src/cluster-test-framework.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# Cluster Test Framework
+
+This document covers the *Cluster Test Framework*.  The CTF is a test harness
+designed for tests that can execute against a local in process cluster and a
```
Comment (garious) [inline book/src/cluster-test-framework.md]: in process -> in-process
and -> or

File: book/src/cluster-test-framework.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+deployed cluster on the network.
```
Comment (garious) [inline book/src/cluster-test-framework.md]: No need for "on the network"

File: book/src/cluster-test-framework.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+
+## Motivation
+
+The goal of the CTF is to provide a framework for writing tests that are stable
+against changes to the cluster implementation.  Regressions can be captured in
```
Comment (garious) [inline book/src/cluster-test-framework.md]: I don't think "against changes to the cluster implementation" describes the importance of this very well. I think you mean something like "independent of where and how the cluster is deployed".  Likewise, when describing regression testing, how a description of a triage process where those devs deploying testnets can handoff tests that are reproducible in both the production deployment and on dev machines?

File: book/src/cluster-test-framework.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+least amount of internal plumbing exposed to the test.
+
+## Design Overview
+
+Tests are provided an entry point, which is a `contact_info::ContactInfo`
+structure, and a keypair that has already been funded.
+
+Each node in the cluster is configured with a `fullnode::FullnodeConfig` at boot
+time.  At boot time this configuration specifies any extra cluster configuration
+required for the test. The cluster, should boot with the configuration when it
```
Comment (garious) [inline book/src/cluster-test-framework.md]: No comma needed after cluster or in-process

File: book/src/cluster-test-framework.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+use solana_sdk::signature::{Keypair, KeypairUtil};
+pub fn test_this_behavior(
+    entry_point_info: &ContactInfo,
+    funding_keypair: &Keypair,
+    num_nodes: usize,
+)
+```
+
+
+## Network Discovery
```
Comment (garious) [inline book/src/cluster-test-framework.md]: Network?

File: book/src/cluster-test-framework.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                num_nodes,
+                10_000,
+                100,
+                &fullnode_config
+                );
+```
+
+## How to design a new test
+
+For example, there is a bug that shows that the network fails when it is flooded
```
Comment (garious) [inline book/src/cluster-test-framework.md]: The network fails or the cluster fails?

File: book/src/cluster-test-framework.md [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+with invalid advertised gossip nodes.  Our gossip library and protocol may
+change, but the network still needs to stay resilient to floods of invalid
```
Comment (garious) [inline book/src/cluster-test-framework.md]: Both the network and cluster need to be resilient. I think you mean cluster though.

File: book/src/cluster-test-framework.md [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+advertised gossip nodes.
+
+Add a new RPCs:
```
Comment (garious) [inline book/src/cluster-test-framework.md]: "Configure the RPC API:"?

File: book/src/SUMMARY.md (hunk 2 above)
Comment (aeyakovenko) [reply book/src/SUMMARY.md]: It’s implemented already.

Comment (garious) [reply book/src/SUMMARY.md]: Grepping "CTF" and "Framework" in the codebase indicates that isn't the case.

Comment (aeyakovenko) [reply book/src/SUMMARY.md]: Rename the LocalCluster to LocalClusterCTF?

Comment (garious) [reply book/src/SUMMARY.md]: No, that wouldn't make any sense.  I would think there'd be a RemoteCluster that launches cloud nodes, and that integration tests would be coded against some `cluster_test_framework.rs`, and that they'd have a `dev-dependency` on a `local_cluster` that could be trivially replaced with a `remote_cluster` dependency to run the same test against a RemoteCluster.

**Comment by aeyakovenko** [comment]
@garious.  Thank you!!!

**Pull request merged**